
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Evaluation of ion suppression correction &#8212; Ion suppression correction  documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/nature.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="index.html">Ion suppression correction  documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Evaluation of ion suppression correction</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars and line breaks on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
    white-space: pre;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Evaluation-of-ion-suppression-correction">
<h1>Evaluation of ion suppression correction<a class="headerlink" href="#Evaluation-of-ion-suppression-correction" title="Permalink to this heading">Â¶</a></h1>
<p>In this notebook, I explore different measures to quantify the effect of correcting SpaceM ion intensity data for partial pixel-cell overlap.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">platform</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">from</span> <span class="nn">src.correction</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">src</span> <span class="kn">import</span> <span class="n">const</span>
<span class="kn">from</span> <span class="nn">sklearn.cluster</span> <span class="kn">import</span> <span class="n">KMeans</span>
<span class="kn">from</span> <span class="nn">importlib</span> <span class="kn">import</span> <span class="n">reload</span>
<span class="c1"># import outer_spacem as osm</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">platform</span><span class="o">.</span><span class="n">system</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;Darwin&quot;</span><span class="p">:</span>
    <span class="n">target_path</span> <span class="o">=</span> <span class="s1">&#39;/Volumes/mklein/FDA_project/data/Mx_Co_Cultured&#39;</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">target_path</span> <span class="o">=</span> <span class="s1">&#39;/home/mklein/FDA_project/data/Mx_Co_Cultured&#39;</span>

<span class="n">condition_name</span> <span class="o">=</span> <span class="s1">&#39;celltype&#39;</span>
</pre></div>
</div>
</div>
<p>The conditions data for the co-culture datasets is established using fluorescence microscopy. The following code is adapted from Martijn and just shows the classification of fluorescence values to celltypes</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target_path</span><span class="p">,</span> <span class="s1">&#39;pipeline_files/batch_sm_matrix.h5ad&#39;</span><span class="p">))</span>
<span class="n">adata_cor</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target_path</span><span class="p">,</span> <span class="s1">&#39;pipeline_files/corrected_batch_sm_matrix.h5ad&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">condition_metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target_path</span><span class="p">,</span><span class="s1">&#39;MORPHnMOL.csv&#39;</span><span class="p">))</span>
<span class="n">condition_metadata</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="n">const</span><span class="o">.</span><span class="n">CELL_PRE</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">condition_metadata</span><span class="o">.</span><span class="n">ObjectNumber</span><span class="p">]</span>
<span class="n">condition_metadata</span><span class="p">[</span><span class="s1">&#39;GFP&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">condition_metadata</span><span class="o">.</span><span class="n">Intensity_MeanIntensity_GFP_quantif</span>
<span class="n">condition_metadata</span><span class="p">[</span><span class="s1">&#39;MCherry&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">condition_metadata</span><span class="o">.</span><span class="n">Intensity_MeanIntensity_mCherry_quantif</span>
<span class="n">condition_metadata</span><span class="p">[</span><span class="s1">&#39;fluorescence_ratio&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">condition_metadata</span><span class="o">.</span><span class="n">GFP</span> <span class="o">/</span> <span class="n">condition_metadata</span><span class="o">.</span><span class="n">MCherry</span><span class="p">)</span>

<span class="c1">#condition_metadata[&#39;celltype&#39;] = &#39;HeLa&#39; if condition_metadata.fluorescence_ratio &lt; 0.8 else &#39;NIH3T3&#39;</span>
<span class="n">condition_metadata</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">condition_metadata</span><span class="o">.</span><span class="n">fluorescence_ratio</span> <span class="o">&lt;</span> <span class="mf">0.8</span><span class="p">,</span> <span class="s1">&#39;HeLa&#39;</span><span class="p">,</span> <span class="s1">&#39;NIH3T3&#39;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">condition_metadata</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>

<span class="n">plot</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">relplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">condition_metadata</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;GFP&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;MCherry&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;celltype&#39;</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">xscale</span><span class="o">=</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">yscale</span><span class="o">=</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;Cell type attributation based on GFP / MCherry fluorescence ratio&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
HeLa      715
NIH3T3    492
Name: celltype, dtype: int64
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;seaborn.axisgrid.FacetGrid at 0x7fbf39c48a60&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/correction_evaluation_5_2.png" src="_images/correction_evaluation_5_2.png" />
</div>
</div>
<p>The co-culture dataset was originally used in the original SpaceM manuscript. The corresponding metadata file <code class="docutils literal notranslate"><span class="pre">MORPHnMOL.csv</span></code> subsets the list of annotated molecules from 104 to 58. Consequently, only these molecules will be explored in the following. Analogously, only the cells present in both the uncorrected and corrected spatiomolecular matrices are kept for the analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">included_cells</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">condition_metadata</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">all_molecules</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span>
<span class="n">included_molecules</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">condition_metadata</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">included_cells</span><span class="p">,</span> <span class="n">all_molecules</span><span class="p">]</span>
<span class="n">adata_cor</span> <span class="o">=</span> <span class="n">adata_cor</span><span class="p">[</span><span class="n">included_cells</span><span class="p">,</span> <span class="n">all_molecules</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_cor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(1207, 58)
(1207, 58)
</pre></div></div>
</div>
<p>In the following, both spatiomolecular matrices are filtered and scaled to zero-mean and unit variance.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">preprocess</span><span class="p">(</span><span class="n">adata</span><span class="p">):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">adata</span>
    <span class="c1"># sc.pp.normalize_total(adata, target_sum=None)</span>
    <span class="c1"># sc.pp.log1p(adata)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="n">preprocess</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">preprocess</span><span class="p">(</span><span class="n">adata_cor</span><span class="p">)</span>


<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata_cor</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(1137, 58)
(1137, 58)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/mklein/.conda/envs/ion_suppression/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:140: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.
  adata.obs[&#39;n_genes&#39;] = number
/home/mklein/.conda/envs/ion_suppression/lib/python3.10/site-packages/scanpy/preprocessing/_simple.py:140: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.
  adata.obs[&#39;n_genes&#39;] = number
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;correction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;uncorrected&#39;</span>
<span class="n">adata_cor</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;correction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;ISM correction&#39;</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">ion</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">:</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(((</span><span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">()[</span><span class="s1">&#39;C37H71O8P&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">adata_cor</span><span class="o">.</span><span class="n">to_df</span><span class="p">()[</span><span class="s1">&#39;C37H71O8P&#39;</span><span class="p">]</span><span class="o">!=</span> <span class="mi">0</span><span class="p">))</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">ion</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">conc_adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">concat</span><span class="p">({</span><span class="s1">&#39;uncorrected&#39;</span><span class="p">:</span> <span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;ISM correction&#39;</span><span class="p">:</span> <span class="n">adata_cor</span><span class="p">},</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;correction&#39;</span><span class="p">,</span> <span class="n">index_unique</span><span class="o">=</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="n">merge</span><span class="o">=</span><span class="s1">&#39;same&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">conc_adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;correction&#39;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;t-test_overestim_var&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/mklein/.conda/envs/ion_suppression/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:417: RuntimeWarning: overflow encountered in expm1
  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (
/home/mklein/.conda/envs/ion_suppression/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:418: RuntimeWarning: overflow encountered in expm1
  self.expm1_func(mean_rest) + 1e-9
/home/mklein/.conda/envs/ion_suppression/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:417: RuntimeWarning: invalid value encountered in divide
  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (
/home/mklein/.conda/envs/ion_suppression/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:417: RuntimeWarning: overflow encountered in expm1
  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (
/home/mklein/.conda/envs/ion_suppression/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:418: RuntimeWarning: overflow encountered in expm1
  self.expm1_func(mean_rest) + 1e-9
/home/mklein/.conda/envs/ion_suppression/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:417: RuntimeWarning: invalid value encountered in divide
  foldchanges = (self.expm1_func(mean_group) + 1e-9) / (
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_ions_dea</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">conc_adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;ISM correction&#39;</span><span class="p">)</span>
<span class="c1"># tops_dea = pd.concat({&#39;strongly_corr&#39;: top_uncor, &#39;hardly_corr&#39;: top_cor})</span>

<span class="n">adata_cor</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_cor</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span>
<span class="n">impact_ions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">all_ions_dea</span><span class="p">,</span> <span class="n">adata_cor</span><span class="o">.</span><span class="n">var</span><span class="p">[[</span><span class="s1">&#39;names&#39;</span><span class="p">,</span> <span class="s1">&#39;correction_using_ion_pool&#39;</span><span class="p">,</span> <span class="s1">&#39;correction_n_iterations&#39;</span><span class="p">,</span> <span class="s1">&#39;correction_n_datapoints&#39;</span><span class="p">,</span> <span class="s1">&#39;correction_quantreg_slope&#39;</span><span class="p">]],</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;names&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>
<span class="n">impact_ions</span> <span class="o">=</span> <span class="n">impact_ions</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;scores&#39;</span><span class="p">)</span>

<span class="n">sign_impact_ions</span> <span class="o">=</span> <span class="n">impact_ions</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">impact_ions</span><span class="p">[</span><span class="s1">&#39;pvals&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.001</span><span class="p">)</span> <span class="o">&amp;</span>
                                   <span class="p">(</span><span class="n">impact_ions</span><span class="p">[</span><span class="s1">&#39;scores&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="o">-</span><span class="mi">2</span><span class="p">)]</span>
<span class="n">impact_ions</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>names</th>
      <th>scores</th>
      <th>logfoldchanges</th>
      <th>pvals</th>
      <th>pvals_adj</th>
      <th>correction_using_ion_pool</th>
      <th>correction_n_iterations</th>
      <th>correction_n_datapoints</th>
      <th>correction_quantreg_slope</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>57</th>
      <td>C18H36O2</td>
      <td>-17.634972</td>
      <td>NaN</td>
      <td>2.161421e-64</td>
      <td>1.253624e-62</td>
      <td>False</td>
      <td>50</td>
      <td>1199</td>
      <td>-0.870044</td>
    </tr>
    <tr>
      <th>56</th>
      <td>C19H37O6P</td>
      <td>-15.897961</td>
      <td>NaN</td>
      <td>2.744945e-53</td>
      <td>7.960342e-52</td>
      <td>False</td>
      <td>55</td>
      <td>1199</td>
      <td>-0.862567</td>
    </tr>
    <tr>
      <th>55</th>
      <td>C21H41O6P</td>
      <td>-15.456485</td>
      <td>NaN</td>
      <td>8.523231e-51</td>
      <td>1.647825e-49</td>
      <td>False</td>
      <td>50</td>
      <td>1199</td>
      <td>-0.842036</td>
    </tr>
    <tr>
      <th>54</th>
      <td>C35H69O8P</td>
      <td>-15.013201</td>
      <td>-223.107941</td>
      <td>6.014576e-48</td>
      <td>8.721135e-47</td>
      <td>False</td>
      <td>53</td>
      <td>1190</td>
      <td>-1.077399</td>
    </tr>
    <tr>
      <th>53</th>
      <td>C21H43O7P</td>
      <td>-14.978187</td>
      <td>-176.138031</td>
      <td>7.998049e-48</td>
      <td>9.277737e-47</td>
      <td>False</td>
      <td>56</td>
      <td>1184</td>
      <td>-1.002771</td>
    </tr>
    <tr>
      <th>52</th>
      <td>C19H39O7P</td>
      <td>-13.888738</td>
      <td>-149.335220</td>
      <td>8.787171e-42</td>
      <td>8.494265e-41</td>
      <td>False</td>
      <td>60</td>
      <td>1182</td>
      <td>-0.895607</td>
    </tr>
    <tr>
      <th>51</th>
      <td>C27H53O12P</td>
      <td>-12.108463</td>
      <td>NaN</td>
      <td>1.292244e-32</td>
      <td>1.070717e-31</td>
      <td>False</td>
      <td>35</td>
      <td>1197</td>
      <td>-0.759358</td>
    </tr>
    <tr>
      <th>50</th>
      <td>C23H48NO7P</td>
      <td>-11.809659</td>
      <td>-153.818237</td>
      <td>3.626930e-31</td>
      <td>2.629524e-30</td>
      <td>False</td>
      <td>49</td>
      <td>1186</td>
      <td>-0.707036</td>
    </tr>
    <tr>
      <th>49</th>
      <td>C6H13O9P</td>
      <td>-10.133627</td>
      <td>NaN</td>
      <td>1.307206e-23</td>
      <td>8.424215e-23</td>
      <td>False</td>
      <td>55</td>
      <td>1198</td>
      <td>-0.677839</td>
    </tr>
    <tr>
      <th>48</th>
      <td>C34H68NO6P</td>
      <td>-9.568450</td>
      <td>-203.360550</td>
      <td>3.305670e-21</td>
      <td>1.917289e-20</td>
      <td>False</td>
      <td>39</td>
      <td>1123</td>
      <td>-1.031471</td>
    </tr>
    <tr>
      <th>47</th>
      <td>C37H71O8P</td>
      <td>-9.400516</td>
      <td>NaN</td>
      <td>1.375299e-20</td>
      <td>7.251577e-20</td>
      <td>False</td>
      <td>83</td>
      <td>1194</td>
      <td>-0.615412</td>
    </tr>
    <tr>
      <th>46</th>
      <td>C21H41O7P</td>
      <td>-9.059540</td>
      <td>-111.638512</td>
      <td>2.872284e-19</td>
      <td>1.388270e-18</td>
      <td>False</td>
      <td>97</td>
      <td>1166</td>
      <td>-0.625741</td>
    </tr>
    <tr>
      <th>45</th>
      <td>C37H74NO8P</td>
      <td>-7.926698</td>
      <td>-70.454964</td>
      <td>3.692087e-15</td>
      <td>1.647239e-14</td>
      <td>False</td>
      <td>56</td>
      <td>1002</td>
      <td>-0.773227</td>
    </tr>
    <tr>
      <th>44</th>
      <td>C6H11O8P</td>
      <td>-6.875738</td>
      <td>NaN</td>
      <td>7.982818e-12</td>
      <td>3.307167e-11</td>
      <td>False</td>
      <td>75</td>
      <td>1199</td>
      <td>-0.553133</td>
    </tr>
    <tr>
      <th>43</th>
      <td>C16H30O2</td>
      <td>-6.649014</td>
      <td>-60.554401</td>
      <td>3.725665e-11</td>
      <td>1.440590e-10</td>
      <td>False</td>
      <td>83</td>
      <td>1144</td>
      <td>-0.493697</td>
    </tr>
    <tr>
      <th>42</th>
      <td>C35H67O8P</td>
      <td>-5.216333</td>
      <td>-57.738529</td>
      <td>1.995202e-07</td>
      <td>7.232609e-07</td>
      <td>False</td>
      <td>106</td>
      <td>1048</td>
      <td>-0.465015</td>
    </tr>
    <tr>
      <th>41</th>
      <td>C36H69O8P</td>
      <td>-4.872649</td>
      <td>-31.599710</td>
      <td>1.181130e-06</td>
      <td>4.029738e-06</td>
      <td>False</td>
      <td>202</td>
      <td>620</td>
      <td>-0.862575</td>
    </tr>
    <tr>
      <th>40</th>
      <td>C25H44NO7P</td>
      <td>-3.744297</td>
      <td>-48.536350</td>
      <td>1.854572e-04</td>
      <td>5.975844e-04</td>
      <td>False</td>
      <td>40</td>
      <td>996</td>
      <td>-0.410651</td>
    </tr>
    <tr>
      <th>39</th>
      <td>C9H19O11P</td>
      <td>-3.449600</td>
      <td>-38.928047</td>
      <td>5.718109e-04</td>
      <td>1.729774e-03</td>
      <td>False</td>
      <td>136</td>
      <td>930</td>
      <td>-0.390134</td>
    </tr>
    <tr>
      <th>38</th>
      <td>C22H34O2</td>
      <td>-3.438163</td>
      <td>-20.568100</td>
      <td>5.964736e-04</td>
      <td>1.729774e-03</td>
      <td>False</td>
      <td>291</td>
      <td>571</td>
      <td>-0.811377</td>
    </tr>
    <tr>
      <th>37</th>
      <td>C22H32O2</td>
      <td>-3.329820</td>
      <td>-24.068752</td>
      <td>8.831721e-04</td>
      <td>2.439237e-03</td>
      <td>False</td>
      <td>45</td>
      <td>643</td>
      <td>-0.759888</td>
    </tr>
    <tr>
      <th>36</th>
      <td>C42H82NO6P</td>
      <td>-3.122862</td>
      <td>-24.066444</td>
      <td>1.814390e-03</td>
      <td>4.783392e-03</td>
      <td>False</td>
      <td>1000</td>
      <td>490</td>
      <td>-0.800185</td>
    </tr>
    <tr>
      <th>35</th>
      <td>C25H50NO7P</td>
      <td>-2.865872</td>
      <td>-15.456901</td>
      <td>4.197410e-03</td>
      <td>1.058477e-02</td>
      <td>False</td>
      <td>69</td>
      <td>482</td>
      <td>-0.813699</td>
    </tr>
    <tr>
      <th>34</th>
      <td>C18H34O2</td>
      <td>-2.805231</td>
      <td>NaN</td>
      <td>5.071052e-03</td>
      <td>1.225504e-02</td>
      <td>False</td>
      <td>290</td>
      <td>1163</td>
      <td>-0.209781</td>
    </tr>
    <tr>
      <th>33</th>
      <td>C39H75O8P</td>
      <td>-2.780143</td>
      <td>-31.414009</td>
      <td>5.478753e-03</td>
      <td>1.271071e-02</td>
      <td>False</td>
      <td>62</td>
      <td>777</td>
      <td>-0.417531</td>
    </tr>
    <tr>
      <th>32</th>
      <td>C39H71O8P</td>
      <td>-2.693429</td>
      <td>-20.024046</td>
      <td>7.124759e-03</td>
      <td>1.589369e-02</td>
      <td>False</td>
      <td>112</td>
      <td>541</td>
      <td>-0.712281</td>
    </tr>
    <tr>
      <th>31</th>
      <td>C23H46NO7P</td>
      <td>-2.599759</td>
      <td>-26.244358</td>
      <td>9.389982e-03</td>
      <td>2.017107e-02</td>
      <td>False</td>
      <td>231</td>
      <td>811</td>
      <td>-0.355845</td>
    </tr>
    <tr>
      <th>30</th>
      <td>C43H73O8P</td>
      <td>-2.493406</td>
      <td>-14.428416</td>
      <td>1.272466e-02</td>
      <td>2.635822e-02</td>
      <td>False</td>
      <td>59</td>
      <td>422</td>
      <td>-0.893290</td>
    </tr>
    <tr>
      <th>29</th>
      <td>C21H39O6P</td>
      <td>-2.342185</td>
      <td>NaN</td>
      <td>1.925748e-02</td>
      <td>3.851495e-02</td>
      <td>False</td>
      <td>106</td>
      <td>1171</td>
      <td>-0.184186</td>
    </tr>
    <tr>
      <th>28</th>
      <td>C43H82NO8P</td>
      <td>-2.303089</td>
      <td>-15.428082</td>
      <td>2.136460e-02</td>
      <td>4.130489e-02</td>
      <td>False</td>
      <td>85</td>
      <td>470</td>
      <td>-0.708251</td>
    </tr>
    <tr>
      <th>27</th>
      <td>C45H80NO8P</td>
      <td>-2.127173</td>
      <td>-14.929408</td>
      <td>3.351401e-02</td>
      <td>6.270362e-02</td>
      <td>False</td>
      <td>51</td>
      <td>497</td>
      <td>-0.730834</td>
    </tr>
    <tr>
      <th>26</th>
      <td>C29H49O12P</td>
      <td>-2.114226</td>
      <td>-12.031621</td>
      <td>3.460583e-02</td>
      <td>6.272306e-02</td>
      <td>False</td>
      <td>59</td>
      <td>391</td>
      <td>-0.770555</td>
    </tr>
    <tr>
      <th>25</th>
      <td>C43H75O8P</td>
      <td>-2.062670</td>
      <td>-11.943058</td>
      <td>3.925866e-02</td>
      <td>6.900006e-02</td>
      <td>False</td>
      <td>70</td>
      <td>395</td>
      <td>-0.787153</td>
    </tr>
    <tr>
      <th>24</th>
      <td>C37H69O8P</td>
      <td>-1.963241</td>
      <td>-23.502100</td>
      <td>4.974048e-02</td>
      <td>8.485141e-02</td>
      <td>False</td>
      <td>54</td>
      <td>785</td>
      <td>-0.283740</td>
    </tr>
    <tr>
      <th>23</th>
      <td>C41H75O8P</td>
      <td>-1.923157</td>
      <td>-14.247446</td>
      <td>5.458591e-02</td>
      <td>9.045666e-02</td>
      <td>False</td>
      <td>44</td>
      <td>482</td>
      <td>-0.584975</td>
    </tr>
    <tr>
      <th>22</th>
      <td>C44H84NO6P</td>
      <td>-1.908280</td>
      <td>-8.137268</td>
      <td>5.648376e-02</td>
      <td>9.100162e-02</td>
      <td>True</td>
      <td>56</td>
      <td>0</td>
      <td>-1.002771</td>
    </tr>
    <tr>
      <th>21</th>
      <td>C38H73O8P</td>
      <td>-1.858610</td>
      <td>-15.489729</td>
      <td>6.321253e-02</td>
      <td>9.908991e-02</td>
      <td>False</td>
      <td>166</td>
      <td>627</td>
      <td>-0.358979</td>
    </tr>
    <tr>
      <th>20</th>
      <td>C44H78NO10P</td>
      <td>-1.811696</td>
      <td>-7.926500</td>
      <td>7.016910e-02</td>
      <td>1.057312e-01</td>
      <td>True</td>
      <td>56</td>
      <td>0</td>
      <td>-1.002771</td>
    </tr>
    <tr>
      <th>19</th>
      <td>C45H82NO8P</td>
      <td>-1.805725</td>
      <td>-7.742727</td>
      <td>7.109513e-02</td>
      <td>1.057312e-01</td>
      <td>False</td>
      <td>44</td>
      <td>294</td>
      <td>-0.902921</td>
    </tr>
    <tr>
      <th>18</th>
      <td>C45H78NO8P</td>
      <td>-1.736991</td>
      <td>-11.557182</td>
      <td>8.252565e-02</td>
      <td>1.196622e-01</td>
      <td>False</td>
      <td>236</td>
      <td>374</td>
      <td>-0.806176</td>
    </tr>
    <tr>
      <th>17</th>
      <td>C40H76NO10P</td>
      <td>-1.677503</td>
      <td>-7.455782</td>
      <td>9.358418e-02</td>
      <td>1.323874e-01</td>
      <td>False</td>
      <td>317</td>
      <td>241</td>
      <td>-0.995944</td>
    </tr>
    <tr>
      <th>16</th>
      <td>C43H76NO7P</td>
      <td>-1.587548</td>
      <td>-10.466653</td>
      <td>1.125289e-01</td>
      <td>1.553970e-01</td>
      <td>False</td>
      <td>77</td>
      <td>374</td>
      <td>-0.738230</td>
    </tr>
    <tr>
      <th>15</th>
      <td>C43H78NO7P</td>
      <td>-1.564367</td>
      <td>-7.591825</td>
      <td>1.178719e-01</td>
      <td>1.589900e-01</td>
      <td>False</td>
      <td>41</td>
      <td>300</td>
      <td>-0.765473</td>
    </tr>
    <tr>
      <th>14</th>
      <td>C39H76NO8P</td>
      <td>-1.479262</td>
      <td>-22.472864</td>
      <td>1.392092e-01</td>
      <td>1.832576e-01</td>
      <td>False</td>
      <td>84</td>
      <td>883</td>
      <td>-0.193073</td>
    </tr>
    <tr>
      <th>13</th>
      <td>C27H51O12P</td>
      <td>-1.468223</td>
      <td>-12.197015</td>
      <td>1.421826e-01</td>
      <td>1.832576e-01</td>
      <td>False</td>
      <td>43</td>
      <td>629</td>
      <td>-0.277594</td>
    </tr>
    <tr>
      <th>12</th>
      <td>C41H73O8P</td>
      <td>-1.450046</td>
      <td>-14.633781</td>
      <td>1.471841e-01</td>
      <td>1.855799e-01</td>
      <td>False</td>
      <td>100</td>
      <td>700</td>
      <td>-0.259073</td>
    </tr>
    <tr>
      <th>11</th>
      <td>C41H74NO7P</td>
      <td>-1.373820</td>
      <td>-8.925296</td>
      <td>1.696338e-01</td>
      <td>2.093353e-01</td>
      <td>False</td>
      <td>149</td>
      <td>365</td>
      <td>-0.608046</td>
    </tr>
    <tr>
      <th>10</th>
      <td>C41H80NO8P</td>
      <td>-1.282632</td>
      <td>-27.822752</td>
      <td>1.997519e-01</td>
      <td>2.413668e-01</td>
      <td>False</td>
      <td>52</td>
      <td>1020</td>
      <td>-0.131633</td>
    </tr>
    <tr>
      <th>9</th>
      <td>C41H74NO8P</td>
      <td>-1.224873</td>
      <td>-6.339025</td>
      <td>2.207508e-01</td>
      <td>2.612969e-01</td>
      <td>False</td>
      <td>71</td>
      <td>301</td>
      <td>-0.576335</td>
    </tr>
    <tr>
      <th>8</th>
      <td>C41H71O8P</td>
      <td>-1.116971</td>
      <td>-9.491037</td>
      <td>2.641250e-01</td>
      <td>3.063850e-01</td>
      <td>False</td>
      <td>74</td>
      <td>626</td>
      <td>-0.244448</td>
    </tr>
    <tr>
      <th>7</th>
      <td>C42H80NO10P</td>
      <td>-0.775144</td>
      <td>-8.334192</td>
      <td>4.383356e-01</td>
      <td>4.984993e-01</td>
      <td>False</td>
      <td>66</td>
      <td>531</td>
      <td>-0.225700</td>
    </tr>
    <tr>
      <th>6</th>
      <td>C41H78NO8P</td>
      <td>-0.356691</td>
      <td>-6.069189</td>
      <td>7.213563e-01</td>
      <td>7.817747e-01</td>
      <td>False</td>
      <td>45</td>
      <td>771</td>
      <td>-0.057405</td>
    </tr>
    <tr>
      <th>5</th>
      <td>C43H76NO8P</td>
      <td>-0.348018</td>
      <td>-4.250483</td>
      <td>7.278592e-01</td>
      <td>7.817747e-01</td>
      <td>False</td>
      <td>51</td>
      <td>660</td>
      <td>-0.071895</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C43H79O13P</td>
      <td>-0.265914</td>
      <td>-2.213482</td>
      <td>7.903293e-01</td>
      <td>8.334382e-01</td>
      <td>False</td>
      <td>40</td>
      <td>343</td>
      <td>-0.115358</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C39H73O8P</td>
      <td>-0.132252</td>
      <td>-3.146961</td>
      <td>8.947966e-01</td>
      <td>9.104948e-01</td>
      <td>False</td>
      <td>58</td>
      <td>962</td>
      <td>-0.014845</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C43H81O13P</td>
      <td>0.023955</td>
      <td>0.754455</td>
      <td>9.808908e-01</td>
      <td>9.808908e-01</td>
      <td>False</td>
      <td>68</td>
      <td>582</td>
      <td>0.005691</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C43H78NO8P</td>
      <td>0.233421</td>
      <td>5.984263</td>
      <td>8.154555e-01</td>
      <td>8.445789e-01</td>
      <td>False</td>
      <td>53</td>
      <td>910</td>
      <td>0.029649</td>
    </tr>
    <tr>
      <th>0</th>
      <td>C20H32O2</td>
      <td>0.655207</td>
      <td>NaN</td>
      <td>5.124008e-01</td>
      <td>5.715240e-01</td>
      <td>False</td>
      <td>54</td>
      <td>1060</td>
      <td>0.060849</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># from bioinfokit import visuz</span>
<span class="c1"># visuz.GeneExpression.volcano(df=impact_ions.replace(np.Inf, np.nan).dropna(), lfc=&#39;logfoldchanges&#39;, pv=&#39;pvals_adj&#39;, show=True)</span>
<span class="c1"># impact_ions.replace(np.Inf, np.nan).dropna().sort_values(&#39;logfoldchanges&#39;)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">matplotlib_venn</span> <span class="kn">import</span> <span class="n">venn2</span>

<span class="n">top</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">impact_ions_fc</span> <span class="o">=</span> <span class="n">impact_ions</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;logfoldchanges&#39;</span><span class="p">)</span>

<span class="n">venn2</span><span class="p">([</span><span class="nb">set</span><span class="p">(</span><span class="n">impact_ions</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
       <span class="nb">set</span><span class="p">(</span><span class="n">impact_ions_fc</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">)],</span>
       <span class="n">set_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;by slope&#39;</span><span class="p">,</span> <span class="s1">&#39;by DEA&#39;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;strongly corrected&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">venn2</span><span class="p">([</span><span class="nb">set</span><span class="p">(</span><span class="n">impact_ions</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="n">top</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
       <span class="nb">set</span><span class="p">(</span><span class="n">impact_ions_fc</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="n">top</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">)],</span>
       <span class="n">set_labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;by slope&#39;</span><span class="p">,</span> <span class="s1">&#39;by DEA&#39;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;hardly corrected&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/correction_evaluation_16_0.png" src="_images/correction_evaluation_16_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/correction_evaluation_16_1.png" src="_images/correction_evaluation_16_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">conc_adata_raw</span> <span class="o">=</span> <span class="n">conc_adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">conc_adata_raw</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">conc_adata_raw</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">X</span>
<span class="n">changed_ions_df</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">obs_df</span><span class="p">(</span><span class="n">conc_adata_raw</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="p">([</span><span class="s1">&#39;correction&#39;</span><span class="p">,</span> <span class="s1">&#39;celltype&#39;</span><span class="p">,</span> <span class="s1">&#39;ObjectNumber&#39;</span><span class="p">]</span><span class="o">+</span><span class="nb">list</span><span class="p">(</span><span class="n">conc_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)))</span>
<span class="n">plot_df</span> <span class="o">=</span> <span class="n">changed_ions_df</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;correction&#39;</span><span class="p">,</span> <span class="s1">&#39;celltype&#39;</span><span class="p">,</span> <span class="s1">&#39;ObjectNumber&#39;</span><span class="p">],</span> <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;ion&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ion&#39;</span><span class="p">,</span> <span class="s1">&#39;celltype&#39;</span><span class="p">,</span> <span class="s1">&#39;ObjectNumber&#39;</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;correction&#39;</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s1">&#39;value&#39;</span><span class="p">)</span>
<span class="n">plot_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">impact_ions</span><span class="p">[</span><span class="s1">&#39;pearson&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">plot_df</span><span class="p">[</span><span class="n">plot_df</span><span class="o">.</span><span class="n">ion</span> <span class="o">==</span> <span class="n">ion</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">()[</span><span class="s1">&#39;uncorrected&#39;</span><span class="p">][</span><span class="s1">&#39;ISM correction&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">ion</span> <span class="ow">in</span> <span class="n">impact_ions</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]]</span>
<span class="n">sns</span><span class="o">.</span><span class="n">pairplot</span><span class="p">(</span><span class="n">impact_ions</span><span class="p">[[</span><span class="s1">&#39;scores&#39;</span><span class="p">,</span> <span class="s1">&#39;logfoldchanges&#39;</span><span class="p">,</span> <span class="s1">&#39;correction_quantreg_slope&#39;</span><span class="p">,</span> <span class="s1">&#39;pearson&#39;</span><span class="p">]])</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/tmp/ipykernel_171/1031242053.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  impact_ions[&#39;pearson&#39;] = [plot_df[plot_df.ion == ion].corr()[&#39;uncorrected&#39;][&#39;ISM correction&#39;] for ion in impact_ions[&#39;names&#39;]]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;seaborn.axisgrid.PairGrid at 0x7fbf314a44f0&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/correction_evaluation_18_2.png" src="_images/correction_evaluation_18_2.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot_df[&#39;ISM correction_intersect&#39;] = plot_df[&#39;ISM correction&#39;] * [ 10 ** adata_cor.var.loc[ion, &#39;correction_quantreg_intersect&#39;] for ion in plot_df[&#39;ion&#39;]]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ions_corr</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">impact_ions</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<span class="n">ions_uncorr</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">impact_ions</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tail</span><span class="p">())</span>
<span class="n">ions</span> <span class="o">=</span> <span class="n">ions_corr</span> <span class="o">+</span> <span class="n">ions_uncorr</span> <span class="c1">#+ [&#39;C27H53O12P&#39;, &#39;C18H36O2&#39;, &#39;C41H80NO8P&#39;, &#39;C39H73O8P&#39;, &#39;C43H81O13P&#39;, &#39;C35H69O8P&#39;]</span>
<span class="c1"># ions = list(set(ions))</span>
<span class="n">plot_df</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;strongly corr&#39;</span> <span class="k">if</span> <span class="p">(</span><span class="n">ion</span> <span class="ow">in</span> <span class="n">ions_corr</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;hardly corr&#39;</span> <span class="k">if</span> <span class="p">(</span><span class="n">ion</span> <span class="ow">in</span> <span class="n">ions_uncorr</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;other&#39;</span> <span class="k">for</span> <span class="n">ion</span> <span class="ow">in</span> <span class="n">plot_df</span><span class="p">[</span><span class="s1">&#39;ion&#39;</span><span class="p">]</span> <span class="p">]</span>

<span class="n">slopes</span> <span class="o">=</span> <span class="p">[</span><span class="n">adata_cor</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ion</span><span class="p">,</span> <span class="s1">&#39;correction_quantreg_slope&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">ion</span> <span class="ow">in</span> <span class="n">ions</span><span class="p">]</span>
<span class="n">intersects</span> <span class="o">=</span> <span class="p">[</span><span class="n">adata_cor</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ion</span><span class="p">,</span> <span class="s1">&#39;correction_quantreg_intersect&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">ion</span> <span class="ow">in</span> <span class="n">ions</span><span class="p">]</span>
<span class="n">pearson</span> <span class="o">=</span> <span class="p">[</span><span class="n">plot_df</span><span class="p">[</span><span class="n">plot_df</span><span class="o">.</span><span class="n">ion</span> <span class="o">==</span> <span class="n">ion</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">()[</span><span class="s1">&#39;uncorrected&#39;</span><span class="p">][</span><span class="s1">&#39;ISM correction&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">ion</span> <span class="ow">in</span> <span class="n">ions</span><span class="p">]</span>
<span class="n">grid</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">FacetGrid</span><span class="p">(</span><span class="n">plot_df</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s1">&#39;ion&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">col_wrap</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">sharex</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">col_order</span><span class="o">=</span><span class="n">ions</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;cividis&#39;</span><span class="p">)</span>
<span class="n">grid</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">,</span> <span class="s1">&#39;uncorrected&#39;</span><span class="p">,</span> <span class="s1">&#39;ISM correction&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">add_legend</span><span class="p">()</span>
<span class="n">grid</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">aspect</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">ax</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">grid</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">flat</span><span class="p">):</span>
    <span class="n">lim</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">ax</span><span class="o">.</span><span class="n">get_xlim</span><span class="p">()[</span><span class="mi">1</span><span class="p">],</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">()[</span><span class="mi">1</span><span class="p">]])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">lim</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">lim</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axline</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span> <span class="n">slope</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.04</span><span class="o">*</span><span class="n">lim</span><span class="p">,</span> <span class="mf">0.8</span><span class="o">*</span><span class="n">lim</span><span class="p">,</span> <span class="s1">&#39;quantreg slope = </span><span class="si">%1.3f</span><span class="se">\n</span><span class="s1">intersect = </span><span class="si">%1.3f</span><span class="se">\n</span><span class="s1">pearson r = </span><span class="si">%1.5f</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">slopes</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">intersects</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">pearson</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/tmp/ipykernel_171/1204562358.py:9: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  pearson = [plot_df[plot_df.ion == ion].corr()[&#39;uncorrected&#39;][&#39;ISM correction&#39;] for ion in ions]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/correction_evaluation_20_1.png" src="_images/correction_evaluation_20_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ea_ad</span> <span class="o">=</span> <span class="n">conc_adata_raw</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">ea_ad</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
<span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target_path</span><span class="p">,</span> <span class="s1">&#39;Mx_coculture_lion_table.csv&#39;</span><span class="p">))</span>
<span class="c1">#sign_impact_ions[&#39;names&#39;]</span>
<span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">({</span><span class="s1">&#39;sample&#39;</span><span class="p">:</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">),</span> <span class="s1">&#39;condition&#39;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">cond</span><span class="p">)},</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;sample&#39;</span><span class="p">)</span>
<span class="n">metadata</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">target_path</span><span class="p">,</span> <span class="s1">&#39;Mx_coculture_metadata.csv&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cond</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">ea_ad</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;correction&#39;</span><span class="p">])</span>
<span class="n">annot</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">ea_ad</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">[</span><span class="n">df</span><span class="p">,</span> <span class="n">annot</span><span class="p">,</span> <span class="n">cond</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">rpy2.robjects</span> <span class="k">as</span> <span class="nn">ro</span>
<span class="kn">from</span> <span class="nn">rpy2.robjects.packages</span> <span class="kn">import</span> <span class="n">importr</span>
<span class="kn">from</span> <span class="nn">rpy2.robjects</span> <span class="kn">import</span> <span class="n">pandas2ri</span>

<span class="kn">from</span> <span class="nn">rpy2.robjects.conversion</span> <span class="kn">import</span> <span class="n">localconverter</span>

<span class="n">bmet</span> <span class="o">=</span> <span class="n">importr</span><span class="p">(</span><span class="s1">&#39;bmetenrichr&#39;</span><span class="p">,</span> <span class="n">lib_loc</span><span class="o">=</span><span class="s1">&#39;/home/mklein/.conda/envs/ion_suppression/lib/R/library&#39;</span><span class="p">)</span>
<span class="n">pbl</span> <span class="o">=</span> <span class="n">importr</span><span class="p">(</span><span class="s1">&#39;pbapply&#39;</span><span class="p">,</span> <span class="n">lib_loc</span><span class="o">=</span><span class="s1">&#39;/home/mklein/.conda/envs/ion_suppression/lib/R/library&#39;</span><span class="p">)</span>
<span class="n">base</span> <span class="o">=</span> <span class="n">importr</span><span class="p">(</span><span class="s1">&#39;base&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">localconverter</span><span class="p">(</span><span class="n">ro</span><span class="o">.</span><span class="n">default_converter</span> <span class="o">+</span> <span class="n">pandas2ri</span><span class="o">.</span><span class="n">converter</span><span class="p">):</span>
    <span class="n">scmatrix</span> <span class="o">=</span> <span class="n">ro</span><span class="o">.</span><span class="n">conversion</span><span class="o">.</span><span class="n">py2rpy</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="n">ro</span><span class="o">.</span><span class="n">conversion</span><span class="o">.</span><span class="n">py2rpy</span><span class="p">(</span><span class="n">annot</span><span class="p">)</span>
    <span class="n">conditions</span> <span class="o">=</span> <span class="n">ro</span><span class="o">.</span><span class="n">conversion</span><span class="o">.</span><span class="n">py2rpy</span><span class="p">(</span><span class="n">cond</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/mklein/.conda/envs/ion_suppression/lib/python3.10/site-packages/rpy2/robjects/pandas2ri.py:54: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for name, values in obj.iteritems():
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">enr</span> <span class="o">=</span> <span class="n">bmet</span><span class="o">.</span><span class="n">initEnrichment</span><span class="p">(</span><span class="n">scmatrix</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">as_matrix</span><span class="p">(</span><span class="n">scmatrix</span><span class="p">),</span>
                 <span class="n">annotations</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">as_character</span><span class="p">(</span><span class="n">annotations</span><span class="p">),</span>
                 <span class="n">conditions</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">as_character</span><span class="p">(</span><span class="n">conditions</span><span class="p">),</span>
                 <span class="n">condition_x</span> <span class="o">=</span> <span class="s2">&quot;uncorrected&quot;</span><span class="p">,</span>
                 <span class="n">condition_y</span> <span class="o">=</span> <span class="s2">&quot;ISM correction&quot;</span>
                <span class="p">)</span>

<span class="n">enr</span> <span class="o">=</span> <span class="n">bmet</span><span class="o">.</span><span class="n">rankScore</span><span class="p">(</span><span class="n">enr</span><span class="p">,</span> <span class="s1">&#39;t.test&#39;</span><span class="p">)</span>
<span class="n">enr</span> <span class="o">=</span> <span class="n">bmet</span><span class="o">.</span><span class="n">calcEnrichment</span><span class="p">(</span><span class="n">enr</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Parsing isomers...
single-cell metabolomics matrix of 58 metabolites and 2274 cells
active pathway: LION

conditions: uncorrected, ISM correction

condition.x: uncorrected
condition.y: ISM correction

Bootstrapping...
  |                                                  | 0 % ~calculating   |+                                                 | 1 % ~01s           |+                                                 | 2 % ~01s           |++                                                | 3 % ~01s           |++                                                | 4 % ~01s           |+++                                               | 5 % ~01s           |+++                                               | 6 % ~01s           |++++                                              | 7 % ~01s           |++++                                              | 8 % ~00s           |+++++                                             | 9 % ~00s           |+++++                                             | 10% ~00s           |++++++                                            | 11% ~00s           |++++++                                            | 12% ~00s           |+++++++                                           | 13% ~00s           |+++++++                                           | 14% ~00s           |++++++++                                          | 15% ~00s           |++++++++                                          | 16% ~00s           |+++++++++                                         | 17% ~00s           |+++++++++                                         | 18% ~00s           |++++++++++                                        | 19% ~00s           |++++++++++                                        | 20% ~00s           |+++++++++++                                       | 21% ~00s           |+++++++++++                                       | 22% ~00s           |++++++++++++                                      | 23% ~00s           |++++++++++++                                      | 24% ~00s           |+++++++++++++                                     | 25% ~00s           |+++++++++++++                                     | 26% ~00s           |++++++++++++++                                    | 27% ~00s           |++++++++++++++                                    | 28% ~00s           |+++++++++++++++                                   | 29% ~00s           |+++++++++++++++                                   | 30% ~00s           |++++++++++++++++                                  | 31% ~00s           |++++++++++++++++                                  | 32% ~00s           |+++++++++++++++++                                 | 33% ~00s           |+++++++++++++++++                                 | 34% ~00s           |++++++++++++++++++                                | 35% ~00s           |++++++++++++++++++                                | 36% ~00s           |+++++++++++++++++++                               | 37% ~00s           |+++++++++++++++++++                               | 38% ~00s           |++++++++++++++++++++                              | 39% ~00s           |++++++++++++++++++++                              | 40% ~00s           |+++++++++++++++++++++                             | 41% ~00s           |+++++++++++++++++++++                             | 42% ~00s           |++++++++++++++++++++++                            | 43% ~00s           |++++++++++++++++++++++                            | 44% ~00s           |+++++++++++++++++++++++                           | 45% ~00s           |+++++++++++++++++++++++                           | 46% ~00s           |++++++++++++++++++++++++                          | 47% ~00s           |++++++++++++++++++++++++                          | 48% ~00s           |+++++++++++++++++++++++++                         | 49% ~00s           |+++++++++++++++++++++++++                         | 50% ~00s           |++++++++++++++++++++++++++                        | 51% ~00s           |++++++++++++++++++++++++++                        | 52% ~00s           |+++++++++++++++++++++++++++                       | 53% ~00s           |+++++++++++++++++++++++++++                       | 54% ~00s           |++++++++++++++++++++++++++++                      | 55% ~00s           |++++++++++++++++++++++++++++                      | 56% ~00s           |+++++++++++++++++++++++++++++                     | 57% ~00s           |+++++++++++++++++++++++++++++                     | 58% ~00s           |++++++++++++++++++++++++++++++                    | 59% ~00s           |++++++++++++++++++++++++++++++                    | 60% ~00s           |+++++++++++++++++++++++++++++++                   | 61% ~00s           |+++++++++++++++++++++++++++++++                   | 62% ~00s           |++++++++++++++++++++++++++++++++                  | 63% ~00s           |++++++++++++++++++++++++++++++++                  | 64% ~00s           |+++++++++++++++++++++++++++++++++                 | 65% ~00s           |+++++++++++++++++++++++++++++++++                 | 66% ~00s           |++++++++++++++++++++++++++++++++++                | 67% ~00s           |++++++++++++++++++++++++++++++++++                | 68% ~00s           |+++++++++++++++++++++++++++++++++++               | 69% ~00s           |+++++++++++++++++++++++++++++++++++               | 70% ~00s           |++++++++++++++++++++++++++++++++++++              | 71% ~00s           |++++++++++++++++++++++++++++++++++++              | 72% ~00s           |+++++++++++++++++++++++++++++++++++++             | 73% ~00s           |+++++++++++++++++++++++++++++++++++++             | 74% ~00s           |++++++++++++++++++++++++++++++++++++++            | 75% ~00s           |++++++++++++++++++++++++++++++++++++++            | 76% ~00s           |+++++++++++++++++++++++++++++++++++++++           | 77% ~00s           |+++++++++++++++++++++++++++++++++++++++           | 78% ~00s           |++++++++++++++++++++++++++++++++++++++++          | 79% ~00s           |++++++++++++++++++++++++++++++++++++++++          | 80% ~00s           |+++++++++++++++++++++++++++++++++++++++++         | 81% ~00s           |+++++++++++++++++++++++++++++++++++++++++         | 82% ~00s           |++++++++++++++++++++++++++++++++++++++++++        | 83% ~00s           |++++++++++++++++++++++++++++++++++++++++++        | 84% ~00s           |+++++++++++++++++++++++++++++++++++++++++++       | 85% ~00s           |+++++++++++++++++++++++++++++++++++++++++++       | 86% ~00s           |++++++++++++++++++++++++++++++++++++++++++++      | 87% ~00s           |++++++++++++++++++++++++++++++++++++++++++++      | 88% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++     | 89% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++     | 90% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++    | 91% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++    | 92% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++++   | 93% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++++   | 94% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++++  | 95% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++++  | 96% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++++++ | 97% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++++++ | 98% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 99% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s

Match to pathway...
  |                                                  | 0 % ~calculating   |+                                                 | 1 % ~15s           |+                                                 | 2 % ~16s           |++                                                | 3 % ~16s           |++                                                | 4 % ~17s           |+++                                               | 5 % ~17s           |+++                                               | 6 % ~17s           |++++                                              | 7 % ~17s           |++++                                              | 8 % ~16s           |+++++                                             | 9 % ~16s           |+++++                                             | 10% ~15s           |++++++                                            | 11% ~15s           |++++++                                            | 12% ~14s           |+++++++                                           | 13% ~14s           |+++++++                                           | 14% ~14s           |++++++++                                          | 15% ~13s           |++++++++                                          | 16% ~13s           |+++++++++                                         | 17% ~13s           |+++++++++                                         | 18% ~12s           |++++++++++                                        | 19% ~12s           |++++++++++                                        | 20% ~12s           |+++++++++++                                       | 21% ~12s           |+++++++++++                                       | 22% ~11s           |++++++++++++                                      | 23% ~11s           |++++++++++++                                      | 24% ~11s           |+++++++++++++                                     | 25% ~11s           |+++++++++++++                                     | 26% ~11s           |++++++++++++++                                    | 27% ~11s           |++++++++++++++                                    | 28% ~10s           |+++++++++++++++                                   | 29% ~10s           |+++++++++++++++                                   | 30% ~10s           |++++++++++++++++                                  | 31% ~10s           |++++++++++++++++                                  | 32% ~09s           |+++++++++++++++++                                 | 33% ~09s           |+++++++++++++++++                                 | 34% ~09s           |++++++++++++++++++                                | 35% ~09s           |++++++++++++++++++                                | 36% ~09s           |+++++++++++++++++++                               | 37% ~09s           |+++++++++++++++++++                               | 38% ~09s           |++++++++++++++++++++                              | 39% ~09s           |++++++++++++++++++++                              | 40% ~09s           |+++++++++++++++++++++                             | 41% ~09s           |+++++++++++++++++++++                             | 42% ~08s           |++++++++++++++++++++++                            | 43% ~08s           |++++++++++++++++++++++                            | 44% ~08s           |+++++++++++++++++++++++                           | 45% ~08s           |+++++++++++++++++++++++                           | 46% ~08s           |++++++++++++++++++++++++                          | 47% ~08s           |++++++++++++++++++++++++                          | 48% ~08s           |+++++++++++++++++++++++++                         | 49% ~08s           |+++++++++++++++++++++++++                         | 50% ~07s           |++++++++++++++++++++++++++                        | 51% ~07s           |++++++++++++++++++++++++++                        | 52% ~07s           |+++++++++++++++++++++++++++                       | 53% ~07s           |+++++++++++++++++++++++++++                       | 54% ~07s           |++++++++++++++++++++++++++++                      | 55% ~07s           |++++++++++++++++++++++++++++                      | 56% ~06s           |+++++++++++++++++++++++++++++                     | 57% ~06s           |+++++++++++++++++++++++++++++                     | 58% ~06s           |++++++++++++++++++++++++++++++                    | 59% ~06s           |++++++++++++++++++++++++++++++                    | 60% ~06s           |+++++++++++++++++++++++++++++++                   | 61% ~06s           |+++++++++++++++++++++++++++++++                   | 62% ~06s           |++++++++++++++++++++++++++++++++                  | 63% ~05s           |++++++++++++++++++++++++++++++++                  | 64% ~05s           |+++++++++++++++++++++++++++++++++                 | 65% ~05s           |+++++++++++++++++++++++++++++++++                 | 66% ~05s           |++++++++++++++++++++++++++++++++++                | 67% ~05s           |++++++++++++++++++++++++++++++++++                | 68% ~05s           |+++++++++++++++++++++++++++++++++++               | 69% ~04s           |+++++++++++++++++++++++++++++++++++               | 70% ~04s           |++++++++++++++++++++++++++++++++++++              | 71% ~04s           |++++++++++++++++++++++++++++++++++++              | 72% ~04s           |+++++++++++++++++++++++++++++++++++++             | 73% ~04s           |+++++++++++++++++++++++++++++++++++++             | 74% ~04s           |++++++++++++++++++++++++++++++++++++++            | 75% ~04s           |++++++++++++++++++++++++++++++++++++++            | 76% ~03s           |+++++++++++++++++++++++++++++++++++++++           | 77% ~03s           |+++++++++++++++++++++++++++++++++++++++           | 78% ~03s           |++++++++++++++++++++++++++++++++++++++++          | 79% ~03s           |++++++++++++++++++++++++++++++++++++++++          | 80% ~03s           |+++++++++++++++++++++++++++++++++++++++++         | 81% ~03s           |+++++++++++++++++++++++++++++++++++++++++         | 82% ~03s           |++++++++++++++++++++++++++++++++++++++++++        | 83% ~02s           |++++++++++++++++++++++++++++++++++++++++++        | 84% ~02s           |+++++++++++++++++++++++++++++++++++++++++++       | 85% ~02s           |+++++++++++++++++++++++++++++++++++++++++++       | 86% ~02s           |++++++++++++++++++++++++++++++++++++++++++++      | 87% ~02s           |++++++++++++++++++++++++++++++++++++++++++++      | 88% ~02s           |+++++++++++++++++++++++++++++++++++++++++++++     | 89% ~02s           |+++++++++++++++++++++++++++++++++++++++++++++     | 90% ~01s           |++++++++++++++++++++++++++++++++++++++++++++++    | 91% ~01s           |++++++++++++++++++++++++++++++++++++++++++++++    | 92% ~01s           |+++++++++++++++++++++++++++++++++++++++++++++++   | 93% ~01s           |+++++++++++++++++++++++++++++++++++++++++++++++   | 94% ~01s           |++++++++++++++++++++++++++++++++++++++++++++++++  | 95% ~01s           |++++++++++++++++++++++++++++++++++++++++++++++++  | 96% ~01s           |+++++++++++++++++++++++++++++++++++++++++++++++++ | 97% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++++++ | 98% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 99% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=14s
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
R[write to console]: 93.1% of annotations were matched to pathway

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Perform enrichment analysis...
  |                                                  | 0 % ~calculating   |+                                                 | 1 % ~05s           |+                                                 | 2 % ~05s           |++                                                | 3 % ~04s           |++                                                | 4 % ~04s           |+++                                               | 5 % ~04s           |+++                                               | 6 % ~05s           |++++                                              | 7 % ~05s           |++++                                              | 8 % ~05s           |+++++                                             | 9 % ~05s           |+++++                                             | 10% ~04s           |++++++                                            | 11% ~05s           |++++++                                            | 12% ~05s           |+++++++                                           | 13% ~04s           |+++++++                                           | 14% ~04s           |++++++++                                          | 15% ~04s           |++++++++                                          | 16% ~04s           |+++++++++                                         | 17% ~04s           |+++++++++                                         | 18% ~04s           |++++++++++                                        | 19% ~04s           |++++++++++                                        | 20% ~04s           |+++++++++++                                       | 21% ~04s           |+++++++++++                                       | 22% ~05s           |++++++++++++                                      | 23% ~05s           |++++++++++++                                      | 24% ~05s           |+++++++++++++                                     | 25% ~05s           |+++++++++++++                                     | 26% ~05s           |++++++++++++++                                    | 27% ~05s           |++++++++++++++                                    | 28% ~05s           |+++++++++++++++                                   | 29% ~04s           |+++++++++++++++                                   | 30% ~04s           |++++++++++++++++                                  | 31% ~04s           |++++++++++++++++                                  | 32% ~04s           |+++++++++++++++++                                 | 33% ~04s           |+++++++++++++++++                                 | 34% ~04s           |++++++++++++++++++                                | 35% ~04s           |++++++++++++++++++                                | 36% ~04s           |+++++++++++++++++++                               | 37% ~04s           |+++++++++++++++++++                               | 38% ~04s           |++++++++++++++++++++                              | 39% ~04s           |++++++++++++++++++++                              | 40% ~03s           |+++++++++++++++++++++                             | 41% ~03s           |+++++++++++++++++++++                             | 42% ~03s           |++++++++++++++++++++++                            | 43% ~03s           |++++++++++++++++++++++                            | 44% ~03s           |+++++++++++++++++++++++                           | 45% ~03s           |+++++++++++++++++++++++                           | 46% ~03s           |++++++++++++++++++++++++                          | 47% ~03s           |++++++++++++++++++++++++                          | 48% ~03s           |+++++++++++++++++++++++++                         | 49% ~03s           |+++++++++++++++++++++++++                         | 50% ~03s           |++++++++++++++++++++++++++                        | 51% ~03s           |++++++++++++++++++++++++++                        | 52% ~03s           |+++++++++++++++++++++++++++                       | 53% ~03s           |+++++++++++++++++++++++++++                       | 54% ~03s           |++++++++++++++++++++++++++++                      | 55% ~02s           |++++++++++++++++++++++++++++                      | 56% ~02s           |+++++++++++++++++++++++++++++                     | 57% ~02s           |+++++++++++++++++++++++++++++                     | 58% ~02s           |++++++++++++++++++++++++++++++                    | 59% ~02s           |++++++++++++++++++++++++++++++                    | 60% ~02s           |+++++++++++++++++++++++++++++++                   | 61% ~02s           |+++++++++++++++++++++++++++++++                   | 62% ~02s           |++++++++++++++++++++++++++++++++                  | 63% ~02s           |++++++++++++++++++++++++++++++++                  | 64% ~02s           |+++++++++++++++++++++++++++++++++                 | 65% ~02s           |+++++++++++++++++++++++++++++++++                 | 66% ~02s           |++++++++++++++++++++++++++++++++++                | 67% ~02s           |++++++++++++++++++++++++++++++++++                | 68% ~02s           |+++++++++++++++++++++++++++++++++++               | 69% ~02s           |+++++++++++++++++++++++++++++++++++               | 70% ~02s           |++++++++++++++++++++++++++++++++++++              | 71% ~02s           |++++++++++++++++++++++++++++++++++++              | 72% ~01s           |+++++++++++++++++++++++++++++++++++++             | 73% ~01s           |+++++++++++++++++++++++++++++++++++++             | 74% ~01s           |++++++++++++++++++++++++++++++++++++++            | 75% ~01s           |++++++++++++++++++++++++++++++++++++++            | 76% ~01s           |+++++++++++++++++++++++++++++++++++++++           | 77% ~01s           |+++++++++++++++++++++++++++++++++++++++           | 78% ~01s           |++++++++++++++++++++++++++++++++++++++++          | 79% ~01s           |++++++++++++++++++++++++++++++++++++++++          | 80% ~01s           |+++++++++++++++++++++++++++++++++++++++++         | 81% ~01s           |+++++++++++++++++++++++++++++++++++++++++         | 82% ~01s           |++++++++++++++++++++++++++++++++++++++++++        | 83% ~01s           |++++++++++++++++++++++++++++++++++++++++++        | 84% ~01s           |+++++++++++++++++++++++++++++++++++++++++++       | 85% ~01s           |+++++++++++++++++++++++++++++++++++++++++++       | 86% ~01s           |++++++++++++++++++++++++++++++++++++++++++++      | 87% ~01s           |++++++++++++++++++++++++++++++++++++++++++++      | 88% ~01s           |+++++++++++++++++++++++++++++++++++++++++++++     | 89% ~01s           |+++++++++++++++++++++++++++++++++++++++++++++     | 90% ~01s           |++++++++++++++++++++++++++++++++++++++++++++++    | 91% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++    | 92% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++++   | 93% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++++   | 94% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++++  | 95% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++++  | 96% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++++++ | 97% ~00s           |+++++++++++++++++++++++++++++++++++++++++++++++++ | 98% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 99% ~00s           |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=05s
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># bmet.plotEnrichment(enr, min_annotations = 1, q_value_cutoff = .2, by_statistic = &quot;ES&quot;)</span>
<span class="n">bmet</span><span class="o">.</span><span class="n">plotEnrichment</span><span class="p">(</span><span class="n">enr</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="s2">&quot;ES&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
R[write to console]: Error in plotEnrichment.bmetenrich(list(scmatrix = c(265.808380126953,  :
  Not enough enriched terms to visualize

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">RRuntimeError</span>                             Traceback (most recent call last)
Cell <span class="ansi-green-fg">In [26], line 2</span>
<span class="ansi-green-intense-fg ansi-bold">      1</span> <span style="color: rgb(95,135,135)"># bmet.plotEnrichment(enr, min_annotations = 1, q_value_cutoff = .2, by_statistic = &#34;ES&#34;)</span>
<span class="ansi-green-fg">----&gt; 2</span> <span class="ansi-yellow-bg">bmet</span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">.</span><span class="ansi-yellow-bg">plotEnrichment</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">enr</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">1</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">0.5</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg" style="color: rgb(175,0,0)">&#34;</span><span class="ansi-yellow-bg" style="color: rgb(175,0,0)">ES</span><span class="ansi-yellow-bg" style="color: rgb(175,0,0)">&#34;</span><span class="ansi-yellow-bg">)</span>

File <span class="ansi-green-fg">~/.conda/envs/ion_suppression/lib/python3.10/site-packages/rpy2/robjects/functions.py:201</span>, in <span class="ansi-cyan-fg">SignatureTranslatedFunction.__call__</span><span class="ansi-blue-fg">(self, *args, **kwargs)</span>
<span class="ansi-green-intense-fg ansi-bold">    199</span>         v <span style="color: rgb(98,98,98)">=</span> kwargs<span style="color: rgb(98,98,98)">.</span>pop(k)
<span class="ansi-green-intense-fg ansi-bold">    200</span>         kwargs[r_k] <span style="color: rgb(98,98,98)">=</span> v
<span class="ansi-green-fg">--&gt; 201</span> <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> (<span class="ansi-yellow-bg" style="color: rgb(0,135,0)">super</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">SignatureTranslatedFunction</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg" style="color: rgb(0,135,0)">self</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    202</span> <span class="ansi-yellow-bg">        </span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">.</span><span class="ansi-yellow-bg" style="color: rgb(0,0,255)">__call__</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">*</span><span class="ansi-yellow-bg">args</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">*</span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">*</span><span class="ansi-yellow-bg">kwargs</span><span class="ansi-yellow-bg">)</span>)

File <span class="ansi-green-fg">~/.conda/envs/ion_suppression/lib/python3.10/site-packages/rpy2/robjects/functions.py:124</span>, in <span class="ansi-cyan-fg">Function.__call__</span><span class="ansi-blue-fg">(self, *args, **kwargs)</span>
<span class="ansi-green-intense-fg ansi-bold">    122</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">else</span>:
<span class="ansi-green-intense-fg ansi-bold">    123</span>         new_kwargs[k] <span style="color: rgb(98,98,98)">=</span> conversion<span style="color: rgb(98,98,98)">.</span>py2rpy(v)
<span class="ansi-green-fg">--&gt; 124</span> res <span style="color: rgb(98,98,98)">=</span> <span class="ansi-yellow-bg" style="color: rgb(0,135,0)">super</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg">Function</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg" style="color: rgb(0,135,0)">self</span><span class="ansi-yellow-bg">)</span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">.</span><span class="ansi-yellow-bg" style="color: rgb(0,0,255)">__call__</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">*</span><span class="ansi-yellow-bg">new_args</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">*</span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">*</span><span class="ansi-yellow-bg">new_kwargs</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-intense-fg ansi-bold">    125</span> res <span style="color: rgb(98,98,98)">=</span> conversion<span style="color: rgb(98,98,98)">.</span>rpy2py(res)
<span class="ansi-green-intense-fg ansi-bold">    126</span> <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> res

File <span class="ansi-green-fg">~/.conda/envs/ion_suppression/lib/python3.10/site-packages/rpy2/rinterface_lib/conversion.py:45</span>, in <span class="ansi-cyan-fg">_cdata_res_to_rinterface.&lt;locals&gt;._</span><span class="ansi-blue-fg">(*args, **kwargs)</span>
<span class="ansi-green-intense-fg ansi-bold">     44</span> <span class="ansi-bold" style="color: rgb(0,135,0)">def</span> <span style="color: rgb(0,0,255)">_</span>(<span style="color: rgb(98,98,98)">*</span>args, <span style="color: rgb(98,98,98)">*</span><span style="color: rgb(98,98,98)">*</span>kwargs):
<span class="ansi-green-fg">---&gt; 45</span>     cdata <span style="color: rgb(98,98,98)">=</span> <span class="ansi-yellow-bg">function</span><span class="ansi-yellow-bg">(</span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">*</span><span class="ansi-yellow-bg">args</span><span class="ansi-yellow-bg">,</span><span class="ansi-yellow-bg"> </span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">*</span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">*</span><span class="ansi-yellow-bg">kwargs</span><span class="ansi-yellow-bg">)</span>
<span class="ansi-green-intense-fg ansi-bold">     46</span>     <span style="color: rgb(95,135,135)"># TODO: test cdata is of the expected CType</span>
<span class="ansi-green-intense-fg ansi-bold">     47</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> _cdata_to_rinterface(cdata)

File <span class="ansi-green-fg">~/.conda/envs/ion_suppression/lib/python3.10/site-packages/rpy2/rinterface.py:810</span>, in <span class="ansi-cyan-fg">SexpClosure.__call__</span><span class="ansi-blue-fg">(self, *args, **kwargs)</span>
<span class="ansi-green-intense-fg ansi-bold">    803</span>     res <span style="color: rgb(98,98,98)">=</span> rmemory<span style="color: rgb(98,98,98)">.</span>protect(
<span class="ansi-green-intense-fg ansi-bold">    804</span>         openrlib<span style="color: rgb(98,98,98)">.</span>rlib<span style="color: rgb(98,98,98)">.</span>R_tryEval(
<span class="ansi-green-intense-fg ansi-bold">    805</span>             call_r,
<span class="ansi-green-intense-fg ansi-bold">    806</span>             call_context<span style="color: rgb(98,98,98)">.</span>__sexp__<span style="color: rgb(98,98,98)">.</span>_cdata,
<span class="ansi-green-intense-fg ansi-bold">    807</span>             error_occured)
<span class="ansi-green-intense-fg ansi-bold">    808</span>     )
<span class="ansi-green-intense-fg ansi-bold">    809</span>     <span class="ansi-bold" style="color: rgb(0,135,0)">if</span> error_occured[<span style="color: rgb(98,98,98)">0</span>]:
<span class="ansi-green-fg">--&gt; 810</span>         <span class="ansi-bold" style="color: rgb(0,135,0)">raise</span> embedded<span style="color: rgb(98,98,98)">.</span>RRuntimeError(_rinterface<span style="color: rgb(98,98,98)">.</span>_geterrmessage())
<span class="ansi-green-intense-fg ansi-bold">    811</span> <span class="ansi-bold" style="color: rgb(0,135,0)">return</span> res

<span class="ansi-red-fg">RRuntimeError</span>: Error in plotEnrichment.bmetenrich(list(scmatrix = c(265.808380126953,  :
  Not enough enriched terms to visualize

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">enr</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<span>ListVector with 15 elements.</span>
<table>
<tbody>

  <tr>
    <th>
    scmatrix
    </th>
    <td>

<span>FloatMatrix with 131892 elements.</span>
<table>
<tbody>
  <tr>

    <td>
    265.808380
    </td>

    <td>
    489.958191
    </td>

    <td>
    870.396118
    </td>

    <td>
    ...
    </td>

    <td>
    7431.572266
    </td>

    <td>
    3101.317139
    </td>

    <td>
    669.905701
    </td>

  </tr>
</tbody>
</table>

    </td>
  </tr>

  <tr>
    <th>
    annotations
    </th>
    <td>

<span>ListVector with 58 elements.</span>
<table>
<tbody>

  <tr>
    <th>
    [no name]
    </th>
    <td>

<span>StrVector with 5 elements.</span>
<table>
<tbody>
  <tr>

    <td>
    '16:1(9...
    </td>

    <td>
    '16:1(6...
    </td>

    <td>
    '16:1(5...
    </td>

    <td>
    '15:1(4...
    </td>

    <td>
    '16:1(4)'
    </td>

  </tr>
</tbody>
</table>

    </td>
  </tr>

  <tr>
    <th>
    [no name]
    </th>
    <td>

<span>StrVector with 10 elements.</span>
<table>
<tbody>
  <tr>

    <td>
    '18:1(6Z)'
    </td>

    <td>
    '18:1(11Z)'
    </td>

    <td>
    '18:1(11E)'
    </td>

    <td>
    ...
    </td>

    <td>
    '18:1(14E)'
    </td>

    <td>
    '18:1(14Z)'
    </td>

    <td>
    '17:1(4)(...
    </td>

  </tr>
</tbody>
</table>

    </td>
  </tr>

  <tr>
    <th>
    [no name]
    </th>
    <td>

<span>StrVector with 0 elements.</span>
<table>
<tbody>
  <tr>

  </tr>
</tbody>
</table>

    </td>
  </tr>

  <tr>
    <th>
    ...
    </th>
    <td>
    ...
    </td>
  </tr>

  <tr>
    <th>
    [no name]
    </th>
    <td>

<span>StrVector with 0 elements.</span>
<table>
<tbody>
  <tr>

  </tr>
</tbody>
</table>

    </td>
  </tr>

  <tr>
    <th>
    [no name]
    </th>
    <td>

<span>StrVector with 0 elements.</span>
<table>
<tbody>
  <tr>

  </tr>
</tbody>
</table>

    </td>
  </tr>

  <tr>
    <th>
    [no name]
    </th>
    <td>

<span>StrVector with 0 elements.</span>
<table>
<tbody>
  <tr>

  </tr>
</tbody>
</table>

    </td>
  </tr>

</tbody>
</table>

    </td>
  </tr>

  <tr>
    <th>
    annotation.weights
    </th>
    <td>
    <rpy2.rinterface_lib.sexp.NULLType object at 0x7f2cd653afc0> [RTYPES.NILSXP]
    </td>
  </tr>

  <tr>
    <th>
    ...
    </th>
    <td>
    ...
    </td>
  </tr>

  <tr>
    <th>
    ranking.by
    </th>
    <td>

<span>StrVector with 1 elements.</span>
<table>
<tbody>
  <tr>

    <td>
    't.test'
    </td>

  </tr>
</tbody>
</table>

    </td>
  </tr>

  <tr>
    <th>
    rankings
    </th>
    <td>

<span>ListVector with 4 elements.</span>
<table>
<tbody>

  <tr>
    <th>
    rank
    </th>
    <td>
    <rpy2.rinterface.IntSexpVector object at 0x7f2cd1182300> [RTYPES.INTSXP]
    </td>
  </tr>

  <tr>
    <th>
    comparison
    </th>
    <td>
    <rpy2.rinterface_lib.sexp.StrSexpVector object at 0x7f2c86ef7d80> [RTYPES.STRSXP]
    </td>
  </tr>

  <tr>
    <th>
    statistic
    </th>
    <td>
    <rpy2.rinterface.FloatSexpVector object at 0x7f2cab579640> [RTYPES.REALSXP]
    </td>
  </tr>

  <tr>
    <th>
    ranking.by
    </th>
    <td>
    <rpy2.rinterface_lib.sexp.StrSexpVector object at 0x7f2cab578780> [RTYPES.STRSXP]
    </td>
  </tr>

</tbody>
</table>

    </td>
  </tr>

  <tr>
    <th>
    enrichment_analysis
    </th>
    <td>

<span>ListVector with 4 elements.</span>
<table>
<tbody>

  <tr>
    <th>
    table
    </th>
    <td>
    <rpy2.rinterface.ListSexpVector object at 0x7f2cab579640> [RTYPES.VECSXP]
    </td>
  </tr>

  <tr>
    <th>
    n
    </th>
    <td>
    <rpy2.rinterface.IntSexpVector object at 0x7f2cab62a480> [RTYPES.INTSXP]
    </td>
  </tr>

  <tr>
    <th>
    fraction_matched_to_pathway
    </th>
    <td>
    <rpy2.rinterface.FloatSexpVector object at 0x7f2cab62b440> [RTYPES.REALSXP]
    </td>
  </tr>

  <tr>
    <th>
    comparison
    </th>
    <td>
    <rpy2.rinterface_lib.sexp.StrSexpVector object at 0x7f2cd1182300> [RTYPES.STRSXP]
    </td>
  </tr>

</tbody>
</table>

    </td>
  </tr>

</tbody>
</table></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">load_ext</span> rpy2.ipython
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-r notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="n">R</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># library(bmetenrichr)</span>

<span class="n">matrix</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="p">[[</span><span class="mi">1</span><span class="p">]]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">matrix</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">])</span>

<span class="n">test</span> <span class="o">&lt;-</span>
  <span class="n">initEnrichment</span><span class="p">(</span><span class="n">scmatrix</span> <span class="o">=</span> <span class="n">data</span><span class="p">[[</span><span class="mi">1</span><span class="p">]],</span>
                 <span class="n">annotations</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">data</span><span class="p">[[</span><span class="mi">2</span><span class="p">]]),</span>
                 <span class="n">conditions</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">data</span><span class="p">[[</span><span class="mi">3</span><span class="p">]]),</span>
                 <span class="n">condition</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="s2">&quot;uncorrected&quot;</span><span class="p">,</span>
                 <span class="n">condition</span><span class="o">.</span><span class="n">y</span> <span class="o">=</span> <span class="s2">&quot;ISM correction&quot;</span>
                <span class="p">)</span>

<span class="c1"># test &lt;- rankScore(test, ranking.by = &#39;t.test&#39;)</span>
<span class="c1"># test &lt;- calcEnrichment(test, n = 100)</span>
<span class="c1"># plotEnrichment(test, min.annotations = 1, q.value.cutoff = .2, by.statistic = &quot;ES&quot;)</span>
</pre></div>
</div>
</div>
<p>In the following graphs, the uncorrected data is shown before the corrected data. Initially, the data is visualized using PCA and UMAP</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">dimred_pca</span><span class="p">(</span><span class="n">adata</span><span class="p">):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_overview</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;celltype&#39;</span><span class="p">)</span>

<span class="n">dimred_pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">dimred_pca</span><span class="p">(</span><span class="n">adata_cor</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">dimred_umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_dist</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;cosine&#39;</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="n">min_dist</span><span class="p">,</span> <span class="n">spread</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">condition_name</span><span class="p">)</span>
    <span class="c1"># f = osm.pl.highlight_scatterplot(</span>
    <span class="c1">#     data = adata,</span>
    <span class="c1">#     obsm_key = &quot;X_umap&quot;,</span>
    <span class="c1">#     hue = condition_name,</span>
    <span class="c1">#     col = condition_name,</span>
    <span class="c1">#     palette = &quot;cividis&quot;,</span>
    <span class="c1">#     height = 5,</span>
    <span class="c1">#     scatter_kwargs = dict(s=10)</span>
    <span class="c1"># )</span>

    <span class="c1"># f.add_legend(markerscale=3)</span>

    <span class="c1"># plt.xticks([])</span>
    <span class="c1"># plt.yticks([])</span>
    <span class="c1"># plt.show()</span>

<span class="n">dimred_umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">dimred_umap</span><span class="p">(</span><span class="n">adata_cor</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<p>There exist several methods for unsupervised clustering, that can be applied to dimensionality-reduced data. Here, I show kMeans clustering, Leiden clustering and a variant of the latter (adapted from the SpaceM manuscript: curated leiden). Both kMeans and Leiden clustering appear to fail at distinguishing the two celltypes as they split the data along the short axis of the UMAP visualization. However, the celltypes are separated along the long axis. The curated Leiden algorithm from Rappez et
al.Â finds smaller Leiden clusters and assigns them to either celltype depending on the nature of the majority of cells. This method of finer granularity does better reproduce the separation between the two celltypes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.metrics.cluster</span> <span class="kn">import</span> <span class="n">completeness_score</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">accuracy_score</span><span class="p">,</span> <span class="n">silhouette_score</span>

<span class="k">def</span> <span class="nf">clustering_methods</span><span class="p">(</span><span class="n">adata</span><span class="p">):</span>
    <span class="n">kmeans</span> <span class="o">=</span> <span class="n">KMeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;kmeans&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kmeans</span><span class="o">.</span><span class="n">labels_</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;leiden_fine&#39;</span><span class="p">)</span>

    <span class="n">leiden</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden_fine&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="n">leiden_curated</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">leiden</span><span class="p">)</span>
    <span class="n">fc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">leiden</span><span class="p">):</span>
        <span class="n">labels</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">fc</span><span class="p">[</span><span class="n">leiden</span> <span class="o">==</span> <span class="n">cluster</span><span class="p">],</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">leiden_curated</span><span class="p">[</span><span class="n">leiden</span> <span class="o">==</span> <span class="n">cluster</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">labels</span><span class="p">[</span><span class="n">counts</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">counts</span><span class="p">)][</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden_curated&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">leiden_curated</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;kmeans&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s1">&#39;leiden_curated&#39;</span><span class="p">,</span> <span class="s1">&#39;celltype&#39;</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Leiden acccuracy score: </span><span class="si">%1.4f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">accuracy_score</span><span class="p">(</span><span class="n">y_true</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">([</span><span class="s1">&#39;HeLa&#39;</span><span class="p">,</span> <span class="s1">&#39;NIH3T3&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">]),</span> <span class="n">y_pred</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">]))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Curated leiden acccuracy score: </span><span class="si">%1.4f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">accuracy_score</span><span class="p">(</span><span class="n">y_true</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">],</span> <span class="n">y_pred</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden_curated&#39;</span><span class="p">]))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;KMeans completeness score: </span><span class="si">%1.4f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">completeness_score</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;kmeans&#39;</span><span class="p">]))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;KMeans silhouette coefficient: </span><span class="si">%1.4f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">silhouette_score</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;kmeans&#39;</span><span class="p">]))</span>

<span class="n">clustering_methods</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">clustering_methods</span><span class="p">(</span><span class="n">adata_cor</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<p>Linear discriminant analysis is a linear classifier that can be trained and applied to the ion intensity data in order to further examine the separation of the different conditions (celltypes). In general, the performance of the classifier depends on the underlying data, the trained model and how they work with each other. Assuming the model itself has equal performance in both cases, higher the accuracy of the classifier can be attributed to a better separation of the data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.discriminant_analysis</span> <span class="kn">import</span> <span class="n">LinearDiscriminantAnalysis</span>
<span class="kn">from</span> <span class="nn">sklearn.model_selection</span> <span class="kn">import</span> <span class="n">RepeatedStratifiedKFold</span><span class="p">,</span> <span class="n">cross_val_score</span>

<span class="k">def</span> <span class="nf">test_classifier</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">predictors</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>

    <span class="n">cv</span> <span class="o">=</span> <span class="n">RepeatedStratifiedKFold</span><span class="p">(</span><span class="n">n_splits</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_repeats</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">scores</span> <span class="o">=</span> <span class="n">cross_val_score</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">predictors</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">scoring</span><span class="o">=</span><span class="s1">&#39;accuracy&#39;</span><span class="p">,</span> <span class="n">cv</span><span class="o">=</span><span class="n">cv</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">())</span>

    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">,</span> <span class="s1">&#39;lda&#39;</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Classification accuracy after 10-fold cross-validation: </span><span class="si">%1.4f</span><span class="s2"> (Â±</span><span class="si">%1.4f</span><span class="s2">)&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">scores</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">scores</span><span class="p">)))</span>

<span class="k">def</span> <span class="nf">LDA</span><span class="p">(</span><span class="n">adata</span><span class="p">):</span>
    <span class="n">predictors</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">LinearDiscriminantAnalysis</span><span class="p">()</span>
    <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">predictors</span><span class="p">,</span> <span class="n">result</span><span class="p">)</span>

    <span class="n">test_classifier</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">predictors</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;lda&#39;</span><span class="p">)</span>

<span class="n">LDA</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">LDA</span><span class="p">(</span><span class="n">adata_cor</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.ensemble</span> <span class="kn">import</span> <span class="n">RandomForestClassifier</span>

<span class="k">def</span> <span class="nf">random_forest</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_depth</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
    <span class="n">predictors</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">RandomForestClassifier</span><span class="p">(</span><span class="n">random_state</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_depth</span><span class="o">=</span><span class="n">max_depth</span><span class="p">)</span>
    <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">predictors</span><span class="p">,</span> <span class="n">result</span><span class="p">)</span>

    <span class="n">test_classifier</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">predictors</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;random_forest&#39;</span><span class="p">)</span>


<span class="n">random_forest</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">random_forest</span><span class="p">(</span><span class="n">adata_cor</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sklearn.neighbors</span> <span class="kn">import</span> <span class="n">KNeighborsClassifier</span><span class="p">,</span> <span class="n">NeighborhoodComponentsAnalysis</span>
<span class="kn">from</span> <span class="nn">sklearn.pipeline</span> <span class="kn">import</span> <span class="n">Pipeline</span>

<span class="k">def</span> <span class="nf">knn_classifier</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span> <span class="o">=</span> <span class="mi">20</span><span class="p">):</span>
    <span class="n">predictors</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span>
    <span class="n">nca</span> <span class="o">=</span> <span class="n">NeighborhoodComponentsAnalysis</span><span class="p">(</span><span class="n">random_state</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">knn</span> <span class="o">=</span> <span class="n">KNeighborsClassifier</span><span class="p">(</span><span class="n">n_neighbors</span> <span class="o">=</span> <span class="n">n_neighbors</span><span class="p">)</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">Pipeline</span><span class="p">([(</span><span class="s1">&#39;nca&#39;</span><span class="p">,</span> <span class="n">nca</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;knn&#39;</span><span class="p">,</span> <span class="n">knn</span><span class="p">)])</span>
    <span class="n">knn</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">predictors</span><span class="p">,</span> <span class="n">result</span><span class="p">)</span>

    <span class="n">test_classifier</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">knn</span><span class="p">,</span> <span class="n">predictors</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;knn&#39;</span><span class="p">)</span>


<span class="n">knn_classifier</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">knn_classifier</span><span class="p">(</span><span class="n">adata_cor</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Finally, Martijn developed an intermixing metric for the separation of different conditions based on the local neighborhood (euclidian distance) of every data point in dimensionality-reduced space. I have extended this single-valued measure to a curve on neighborhood-scale and introduced a correction factor for unbalanced datasets. For the co-culture dataset, this metric shows a clear difference in the group intermixing of uncorrected and corrected ion intensities. Intermixing is almost 50%
lower after correction across neighborhood sizes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">src.evaluation</span> <span class="kn">import</span> <span class="n">intermixing</span><span class="p">,</span> <span class="n">intermixing_metric_sampled</span>
<span class="n">intermixing</span><span class="p">({</span><span class="s1">&#39;uncorrected&#39;</span><span class="p">:</span> <span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;ISM correction&#39;</span><span class="p">:</span> <span class="n">adata_cor</span><span class="p">},</span> <span class="n">condition_name</span><span class="o">=</span><span class="n">condition_name</span><span class="p">,</span> <span class="n">sample_frac</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>For a better comparison to Martijns results, I also calculate the intermixing metric for the fixed neighborhood of <span class="math notranslate nohighlight">\(n=10\)</span>, with and without normalization to group imbalance. Martijn reports intermixing values of <span class="math notranslate nohighlight">\(0.152\)</span> for ISM correction and <span class="math notranslate nohighlight">\(0.215\)</span> for uncorrected data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">intermixing_metric_sampled</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">condition_name</span><span class="p">,</span> <span class="n">neighborhood_size</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">],</span> <span class="n">n_datapoints</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
<span class="n">intermixing_metric_sampled</span><span class="p">(</span><span class="n">adata_cor</span><span class="p">,</span> <span class="n">condition_name</span><span class="p">,</span> <span class="n">neighborhood_size</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">],</span> <span class="n">n_datapoints</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;ISM&#39;</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
<span class="n">intermixing_metric_sampled</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">condition_name</span><span class="p">,</span> <span class="n">neighborhood_size</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">],</span> <span class="n">n_datapoints</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span>
<span class="n">intermixing_metric_sampled</span><span class="p">(</span><span class="n">adata_cor</span><span class="p">,</span> <span class="n">condition_name</span><span class="p">,</span> <span class="n">neighborhood_size</span><span class="o">=</span><span class="p">[</span><span class="mi">10</span><span class="p">],</span> <span class="n">n_datapoints</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;ISM&#39;</span><span class="p">,</span> <span class="n">normalized</span><span class="o">=</span><span class="kc">False</span><span class="p">)],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="s1">&#39;ISM&#39;</span><span class="p">,</span> <span class="s1">&#39;none_raw&#39;</span><span class="p">,</span> <span class="s1">&#39;ISM_raw&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>As visible in UMAP space, the two celltypes are separated along the long axis of the cluster instead of the short axis. This may indicate that the metabolic profiles within two populations themselves are more heterogenic than between them. Thus an additional biological or technical factor seems to influence the metabolic profiles. To investigate this furter, I did a differential expression analysis between the two leiden clusters. On top of that, it would be useful to complement this with an
enrichment analysis.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;t-test_overestim_var&#39;</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>
<span class="c1"># sc.pl.rank_genes_groups_tracksplot(adata, groupby=&#39;leiden&#39;)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups_violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="s1">&#39;0&#39;</span><span class="p">)[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;data/Mx_Co_Cultured/metabolite-set.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/correction_evaluation.ipynb.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="index.html">Ion suppression correction  documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Evaluation of ion suppression correction</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2022, Marius Klein.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 5.3.0.
    </div>
  </body>
</html>